File: app/src/main/java/org/wikipedia/feed/FeedCoordinatorBase.java
package org.wikipedia.feed;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.feed.accessibility.AccessibilityCard;
import org.wikipedia.feed.aggregated.AggregatedFeedContentClient;
import org.wikipedia.feed.announcement.AnnouncementClient;
import org.wikipedia.feed.becauseyouread.BecauseYouReadClient;
import org.wikipedia.feed.dataclient.FeedClient;
import org.wikipedia.feed.dayheader.DayHeaderCard;
import org.wikipedia.feed.featured.FeaturedArticleCard;
import org.wikipedia.feed.image.FeaturedImageCard;
import org.wikipedia.feed.model.Card;
import org.wikipedia.feed.model.CardType;
import org.wikipedia.feed.mostread.MostReadListCard;
import org.wikipedia.feed.news.NewsCard;
import org.wikipedia.feed.offline.OfflineCard;
import org.wikipedia.feed.onthisday.OnThisDayCard;
import org.wikipedia.feed.progress.ProgressCard;
import org.wikipedia.feed.suggestededits.SuggestedEditsFeedClient;
import org.wikipedia.settings.Prefs;
import org.wikipedia.util.DeviceUtil;
import org.wikipedia.util.ThrowableUtil;
import org.wikipedia.util.log.L;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public abstract class FeedCoordinatorBase {
    private static final int MAX_HIDDEN_CARDS = 100;

    public interface FeedUpdateListener {
        void insert(Card card, int pos);
        void remove(Card card, int pos);
        void finished(boolean shouldUpdatePreviousCard);
    }

    @NonNull private Context context;
    @Nullable private WikiSite wiki;
    @Nullable private FeedUpdateListener updateListener;
    @NonNull private final List<Card> cards = new ArrayList<>();
    private int currentAge;
    private List<FeedClient> pendingClients = new ArrayList<>();
    private FeedClient.Callback callback = new ClientRequestCallback();
    private Card progressCard = new ProgressCard();
    private int currentDayCardAge = -1;

    private Set<String> hiddenCards = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
        @Override
        public boolean removeEldestEntry(Map.Entry<String, Boolean> eldest) {
            return size() > MAX_HIDDEN_CARDS;
        }
    });

    public FeedCoordinatorBase(@NonNull Context context) {
        this.context = context;
        updateHiddenCards();
    }

    public void updateHiddenCards() {
        hiddenCards.clear();
        hiddenCards.addAll(Prefs.getHiddenCards());
    }

    @NonNull
    public List<Card> getCards() {
        return cards;
    }

    public void setFeedUpdateListener(@Nullable FeedUpdateListener listener) {
        updateListener = listener;
    }

    public void reset() {
        wiki = null;
        currentAge = 0;
        currentDayCardAge = -1;
        for (FeedClient client : pendingClients) {
            client.cancel();
        }
        pendingClients.clear();
        cards.clear();
    }

    public void incrementAge() {
        currentAge++;
    }

    public void more(@NonNull WikiSite wiki) {
        this.wiki = wiki;

        if (cards.size() == 0) {
            requestProgressCard();
        }

        if (DeviceUtil.isAccessibilityEnabled()) {
            removeAccessibilityCard();
        }

        buildScript(currentAge);
        requestCard(wiki);
    }

    public boolean finished() {
        return pendingClients.isEmpty();
    }

    public int getAge() {
        return currentAge;
    }

    public int dismissCard(@NonNull Card card) {
        int position = cards.indexOf(card);
        if (card.type() == CardType.RANDOM) {
            FeedContentType.RANDOM.setEnabled(false);
            FeedContentType.saveState();
        } else if (card.type() == CardType.MAIN_PAGE) {
            FeedContentType.MAIN_PAGE.setEnabled(false);
            FeedContentType.saveState();
        } else {
            addHiddenCard(card);
        }
        removeCard(card, position);
        card.onDismiss();
        return position;
    }

    public void undoDismissCard(@NonNull Card card, int position) {
        if (card.type() == CardType.RANDOM) {
            FeedContentType.RANDOM.setEnabled(true);
            FeedContentType.saveState();
        } else if (card.type() == CardType.MAIN_PAGE) {
            FeedContentType.MAIN_PAGE.setEnabled(true);
            FeedContentType.saveState();
        } else {
            unHideCard(card);
        }
        insertCard(card, position);
        card.onRestore();
    }

    protected abstract void buildScript(int age);

    void addPendingClient(@Nullable FeedClient client) {
        if (client != null) {
            pendingClients.add(client);
        }
    }

    void conditionallyAddPendingClient(@Nullable FeedClient client, boolean condition) {
        if (condition && client != null) {
            pendingClients.add(client);
        }
    }

    // Call to kick off the request chain or to retry a failed request.  To move to the next pending
    // client, call requestNextCard.
    private void requestCard(@NonNull WikiSite wiki) {
        if (pendingClients.isEmpty()) {
            removeProgressCard();
            return;
        }
        pendingClients.get(0).request(context, wiki, currentAge, callback);
    }

    private void requestNextCard(@NonNull WikiSite wiki) {
        if (!pendingClients.isEmpty()) {
            pendingClients.remove(0);
        }
        if (!(getLastCard() instanceof ProgressCard) && shouldShowProgressCard(pendingClients.get(0))) {
            requestProgressCard();
        }
        requestCard(wiki);
    }

    void requestOfflineCard() {
        if (!(getLastCard() instanceof OfflineCard)) {
            appendCard(new OfflineCard());
        }
    }

    void removeOfflineCard() {
        if (getLastCard() instanceof OfflineCard) {
            dismissCard(getLastCard());
        }
    }

    private Card getLastCard() {
        return cards.size() > 1 ? cards.get(cards.size() - 1) : null;
    }

    private void requestProgressCard() {
        if (!(getLastCard() instanceof ProgressCard)) {
            appendCard(progressCard);
        }
    }

    private void removeProgressCard() {
        removeCard(progressCard, cards.indexOf(progressCard));
    }

    private void setOfflineState() {
        removeProgressCard();
        appendCard(new OfflineCard());
    }

    private void removeAccessibilityCard() {
        if (getLastCard() instanceof AccessibilityCard) {
            removeCard(getLastCard(), cards.indexOf(getLastCard()));
            getLastCard().onDismiss();
            // TODO: possible on optimization if automatically scroll up to the next card.
        }
    }

    private class ClientRequestCallback implements FeedClient.Callback {
        @Override public void success(@NonNull List<? extends Card> cardList) {
            boolean atLeastOneAppended = false;
            for (Card card : cardList) {
                if (!isCardHidden(card)) {
                    appendCard(card);
                    atLeastOneAppended = true;
                }
            }
            //noinspection ConstantConditions
            requestNextCard(wiki);
            if (pendingClients.isEmpty() && updateListener != null) {
                updateListener.finished(!atLeastOneAppended);
            }
        }

        @Override public void error(@NonNull Throwable caught) {
            if (ThrowableUtil.isOffline(caught)) {
                setOfflineState();
            } else {
                //noinspection ConstantConditions
                requestNextCard(wiki);
                L.w(caught);
            }
        }
    }

    private void appendCard(@NonNull Card card) {
        int progressPos = cards.indexOf(progressCard);
        int pos = progressPos >= 0 ? progressPos : cards.size();

        if (isDailyCardType(card) && currentDayCardAge < currentAge) {
            currentDayCardAge = currentAge;
            insertCard(new DayHeaderCard(currentDayCardAge), pos++);
        }

        insertCard(card, pos);
    }

    private void insertCard(@NonNull Card card, int position) {
        if (position < 0) {
            return;
        }
        cards.add(position, card);
        if (updateListener != null) {
            updateListener.insert(card, position);
        }
    }

    private void removeCard(@NonNull Card card, int position) {
        if (position < 0) {
            return;
        }
        cards.remove(card);
        if (updateListener != null) {
            updateListener.remove(card, position);
        }
    }

    private void addHiddenCard(@NonNull Card card) {
        hiddenCards.add(card.getHideKey());
        Prefs.setHiddenCards(hiddenCards);
    }

    private boolean isCardHidden(@NonNull Card card) {
        return hiddenCards.contains(card.getHideKey());
    }

    private void unHideCard(@NonNull Card card) {
        hiddenCards.remove(card.getHideKey());
        Prefs.setHiddenCards(hiddenCards);
    }

    private boolean isDailyCardType(@NonNull Card card) {
        return card instanceof NewsCard || card instanceof OnThisDayCard
                || card instanceof MostReadListCard || card instanceof FeaturedArticleCard
                || card instanceof FeaturedImageCard;
    }

    private boolean shouldShowProgressCard(@NonNull FeedClient pendingClient) {
        return pendingClient instanceof SuggestedEditsFeedClient
                || pendingClient instanceof AnnouncementClient
                || pendingClient instanceof BecauseYouReadClient;
    }
}

File: app/src/main/java/org/wikipedia/feed/FeedCoordinatorBase.kt
package org.wikipedia.feed

import android.content.Context
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.feed.accessibility.AccessibilityCard
import org.wikipedia.feed.announcement.AnnouncementClient
import org.wikipedia.feed.becauseyouread.BecauseYouReadClient
import org.wikipedia.feed.dataclient.FeedClient
import org.wikipedia.feed.dayheader.DayHeaderCard
import org.wikipedia.feed.featured.FeaturedArticleCard
import org.wikipedia.feed.image.FeaturedImageCard
import org.wikipedia.feed.model.Card
import org.wikipedia.feed.model.CardType
import org.wikipedia.feed.mostread.MostReadListCard
import org.wikipedia.feed.news.NewsCard
import org.wikipedia.feed.offline.OfflineCard
import org.wikipedia.feed.onthisday.OnThisDayCard
import org.wikipedia.feed.progress.ProgressCard
import org.wikipedia.feed.suggestededits.SuggestedEditsFeedClient
import org.wikipedia.settings.Prefs
import org.wikipedia.util.DeviceUtil
import org.wikipedia.util.ThrowableUtil
import org.wikipedia.util.log.L
import java.util.*

abstract class FeedCoordinatorBase(private val context: Context) {

    interface FeedUpdateListener {
        fun insert(card: Card, pos: Int)
        fun remove(card: Card, pos: Int)
        fun finished(shouldUpdatePreviousCard: Boolean)
    }

    private var wiki: WikiSite? = null
    private var updateListener: FeedUpdateListener? = null
    private val pendingClients = mutableListOf<FeedClient>()
    private val callback = ClientRequestCallback()
    private val progressCard = ProgressCard()
    private var currentDayCardAge = -1
    private val hiddenCards =
        Collections.newSetFromMap(object : LinkedHashMap<String?, Boolean?>() {
            public override fun removeEldestEntry(eldest: Map.Entry<String?, Boolean?>): Boolean {
                return size > MAX_HIDDEN_CARDS
            }
        })
    var age = 0
    val cards = mutableListOf<Card>()

    init {
        updateHiddenCards()
    }

    fun updateHiddenCards() {
        hiddenCards.clear()
        hiddenCards.addAll(Prefs.getHiddenCards())
    }

    fun setFeedUpdateListener(listener: FeedUpdateListener?) {
        updateListener = listener
    }

    open fun reset() {
        wiki = null
        age = 0
        currentDayCardAge = -1
        for (client in pendingClients) {
            client.cancel()
        }
        pendingClients.clear()
        cards.clear()
    }

    fun incrementAge() {
        age++
    }

    fun more(wiki: WikiSite) {
        this.wiki = wiki
        if (cards.size == 0) {
            requestProgressCard()
        }
        if (DeviceUtil.isAccessibilityEnabled) {
            removeAccessibilityCard()
        }
        buildScript(age)
        requestCard(wiki)
    }

    fun finished(): Boolean {
        return pendingClients.isEmpty()
    }

    fun dismissCard(card: Card): Int {
        val position = cards.indexOf(card)
        when {
            card.type() === CardType.RANDOM -> {
                FeedContentType.RANDOM.isEnabled = false
                FeedContentType.saveState()
            }
            card.type() === CardType.MAIN_PAGE -> {
                FeedContentType.MAIN_PAGE.isEnabled = false
                FeedContentType.saveState()
            }
            else -> {
                addHiddenCard(card)
            }
        }
        removeCard(card, position)
        card.onDismiss()
        return position
    }

    fun undoDismissCard(card: Card, position: Int) {
        when {
            card.type() === CardType.RANDOM -> {
                FeedContentType.RANDOM.isEnabled = true
                FeedContentType.saveState()
            }
            card.type() === CardType.MAIN_PAGE -> {
                FeedContentType.MAIN_PAGE.isEnabled = true
                FeedContentType.saveState()
            }
            else -> unHideCard(card)
        }
        insertCard(card, position)
        card.onRestore()
    }

    protected abstract fun buildScript(age: Int)
    fun addPendingClient(client: FeedClient?) {
        if (client != null) {
            pendingClients.add(client)
        }
    }

    fun conditionallyAddPendingClient(client: FeedClient?, condition: Boolean) {
        if (condition && client != null) {
            pendingClients.add(client)
        }
    }

    // Call to kick off the request chain or to retry a failed request.  To move to the next pending
    // client, call requestNextCard.
    private fun requestCard(wiki: WikiSite) {
        if (pendingClients.isEmpty()) {
            removeProgressCard()
            return
        }
        pendingClients[0].request(context, wiki, age, callback)
    }

    private fun requestNextCard(wiki: WikiSite) {
        if (pendingClients.isNotEmpty()) {
            pendingClients.removeAt(0)
        }
        if (lastCard !is ProgressCard && shouldShowProgressCard(pendingClients[0])) {
            requestProgressCard()
        }
        requestCard(wiki)
    }

    fun requestOfflineCard() {
        if (lastCard !is OfflineCard) {
            appendCard(OfflineCard())
        }
    }

    fun removeOfflineCard() {
        if (lastCard is OfflineCard) {
            dismissCard(lastCard as OfflineCard)
        }
    }

    private val lastCard get() = if (cards.size > 1) cards[cards.size - 1] else null

    private fun requestProgressCard() {
        if (lastCard !is ProgressCard) {
            appendCard(progressCard)
        }
    }

    private fun removeProgressCard() {
        removeCard(progressCard, cards.indexOf(progressCard))
    }

    private fun setOfflineState() {
        removeProgressCard()
        appendCard(OfflineCard())
    }

    private fun removeAccessibilityCard() {
        if (lastCard is AccessibilityCard) {
            removeCard(lastCard as AccessibilityCard, cards.indexOf(lastCard))
            (lastCard as AccessibilityCard).onDismiss()
            // TODO: possible on optimization if automatically scroll up to the next card.
        }
    }

    private inner class ClientRequestCallback : FeedClient.Callback {
        override fun success(cardList: List<Card>) {
            var atLeastOneAppended = false
            for (card in cardList) {
                if (!isCardHidden(card)) {
                    appendCard(card)
                    atLeastOneAppended = true
                }
            }
            wiki?.let { requestNextCard(it) }
            if (pendingClients.isEmpty()) {
                updateListener?.finished(!atLeastOneAppended)
            }
        }

        override fun error(caught: Throwable) {
            if (ThrowableUtil.isOffline(caught)) {
                setOfflineState()
            } else {
                wiki?.let { requestNextCard(it) }
                L.w(caught)
            }
        }
    }

    private fun appendCard(card: Card) {
        val progressPos = cards.indexOf(progressCard)
        var pos = if (progressPos >= 0) progressPos else cards.size
        if (isDailyCardType(card) && currentDayCardAge < age) {
            currentDayCardAge = age
            insertCard(DayHeaderCard(currentDayCardAge), pos++)
        }
        insertCard(card, pos)
    }

    private fun insertCard(card: Card, position: Int) {
        if (position < 0) {
            return
        }
        cards.add(position, card)
        updateListener?.insert(card, position)
    }

    private fun removeCard(card: Card, position: Int) {
        if (position < 0) {
            return
        }
        cards.remove(card)
        updateListener?.remove(card, position)
    }

    private fun addHiddenCard(card: Card) {
        hiddenCards.add(card.hideKey)
        Prefs.setHiddenCards(hiddenCards)
    }

    private fun isCardHidden(card: Card): Boolean {
        return hiddenCards.contains(card.hideKey)
    }

    private fun unHideCard(card: Card) {
        hiddenCards.remove(card.hideKey)
        Prefs.setHiddenCards(hiddenCards)
    }

    private fun isDailyCardType(card: Card): Boolean {
        return card is NewsCard || card is OnThisDayCard ||
                card is MostReadListCard || card is FeaturedArticleCard ||
                card is FeaturedImageCard
    }

    private fun shouldShowProgressCard(pendingClient: FeedClient): Boolean {
        return pendingClient is SuggestedEditsFeedClient ||
                pendingClient is AnnouncementClient ||
                pendingClient is BecauseYouReadClient
    }

    companion object {
        private const val MAX_HIDDEN_CARDS = 100
    }
}
